Author: CusDeb OS Team
Description: Add DrawShadowText support to comctl32. Implement the full ReactOS
 DrawShadowText algorithm, use it for desktop listview labels when
 LVS_EX_TRANSPARENTSHADOWTEXT is set, and draw taskbar toolbar button labels with a
 theme-defined text colour and a 1px black shadow.
Last-Modified: Mon, 13 Jul 2026 20:00:00 +0300

--- a/src/dlls/comctl32/commctrl.c
+++ b/src/dlls/comctl32/commctrl.c
@@ -1655,35 +1655,115 @@
 /***********************************************************************
  * DrawShadowText [COMCTL32.@]
  *
- * Draw text with shadow.
+ * Draw text with shadow. Ported from ReactOS.
  */
-int WINAPI DrawShadowText(HDC hdc, LPCWSTR text, UINT length, RECT *rect, DWORD flags,
-                          COLORREF crText, COLORREF crShadow, int offset_x, int offset_y)
+int WINAPI DrawShadowText(HDC hdc, LPCWSTR pszText, UINT cch, RECT *prc, DWORD dwFlags,
+                          COLORREF crText, COLORREF crShadow, int ixOffset, int iyOffset)
 {
-    int bkmode, ret;
-    COLORREF clr;
-    RECT r;
+    COLORREF crOldText;
+    RECT rcText;
+    INT iRet, x, y, x2, y2;
+    BYTE *pBits;
+    HBITMAP hbm, hbmOld;
+    BITMAPINFO bi;
+    HDC hdcMem;
+    HFONT hOldFont;
+    BLENDFUNCTION bf;
+
+    /* Create 32 bit DIB section for the shadow */
+    ZeroMemory(&bi, sizeof(bi));
+    bi.bmiHeader.biSize = sizeof(bi.bmiHeader);
+    bi.bmiHeader.biWidth = prc->right - prc->left + 4;
+    bi.bmiHeader.biHeight = prc->bottom - prc->top + 5; /* bottom-up DIB */
+    bi.bmiHeader.biPlanes = 1;
+    bi.bmiHeader.biBitCount = 32;
+    bi.bmiHeader.biCompression = BI_RGB;
+    hbm = CreateDIBSection(hdc, &bi, DIB_RGB_COLORS, (PVOID*)&pBits, NULL, 0);
+    if(!hbm)
+    {
+        ERR("CreateDIBSection failed\n");
+        return 0;
+    }
+
+    /* Create memory device context for new DIB section and select it */
+    hdcMem = CreateCompatibleDC(hdc);
+    if(!hdcMem)
+    {
+        ERR("CreateCompatibleDC failed\n");
+        DeleteObject(hbm);
+        return 0;
+    }
+
+    hbmOld = (HBITMAP)SelectObject(hdcMem, hbm);
+
+    /* Draw text on our helper bitmap */
+    hOldFont = (HFONT)SelectObject(hdcMem, GetCurrentObject(hdc, OBJ_FONT));
+    SetTextColor(hdcMem, RGB(16, 16, 16));
+    SetBkColor(hdcMem, RGB(0, 0, 0));
+    SetBkMode(hdcMem, TRANSPARENT);
+    SetRect(&rcText, 0, 0, prc->right - prc->left, prc->bottom - prc->top);
+    DrawTextW(hdcMem, pszText, cch, &rcText, dwFlags);
+    SelectObject(hdcMem, hOldFont);
+
+    /* Flush GDI so data pointed by pBits is valid */
+    GdiFlush();
+
+    /* Set alpha of pixels (forget about colors for now. They will be changed in next loop).
+       We copy text image 4*5 times and each time alpha is added */
+    for (x = 0; x < bi.bmiHeader.biWidth; ++x)
+        for (y = 0; y < bi.bmiHeader.biHeight; ++y)
+        {
+            BYTE *pDest = &pBits[(y * bi.bmiHeader.biWidth + x) * 4];
+            UINT Alpha = 0;
+
+            for (x2 = x - 4 + 1; x2 <= x; ++x2)
+                for (y2 = y; y2 < y + 5; ++y2)
+                {
+                    if (x2 >= 0 && x2 < bi.bmiHeader.biWidth && y2 >= 0 && y2 < bi.bmiHeader.biHeight)
+                    {
+                        BYTE *pSrc = &pBits[(y2 * bi.bmiHeader.biWidth + x2) * 4];
+                        Alpha += pSrc[0];
+                    }
+                }
+
+            if (Alpha > 255)
+                Alpha = 255;
+            pDest[3] = Alpha;
+        }
+
+    /* Now set the color of each pixel to shadow color * alpha (see GdiAlphaBlend) */
+    for (x = 0; x < bi.bmiHeader.biWidth; ++x)
+        for (y = 0; y < bi.bmiHeader.biHeight; ++y)
+        {
+            BYTE *pDest = &pBits[(y * bi.bmiHeader.biWidth + x) * 4];
+            pDest[0] = GetBValue(crShadow) * pDest[3] / 255;
+            pDest[1] = GetGValue(crShadow) * pDest[3] / 255;
+            pDest[2] = GetRValue(crShadow) * pDest[3] / 255;
+        }
+
+    /* Fix ixOffset of the shadow (tested on Win) */
+    ixOffset -= 3;
+    iyOffset -= 3;
+
+    /* Alpha blend helper image to destination DC */
+    bf.BlendOp = AC_SRC_OVER;
+    bf.BlendFlags = 0;
+    bf.SourceConstantAlpha = 255;
+    bf.AlphaFormat = AC_SRC_ALPHA;
+    GdiAlphaBlend(hdc, prc->left + ixOffset, prc->top + iyOffset, bi.bmiHeader.biWidth, bi.bmiHeader.biHeight, hdcMem, 0, 0, bi.bmiHeader.biWidth, bi.bmiHeader.biHeight, bf);
+
+    /* Delete the helper bitmap */
+    SelectObject(hdcMem, hbmOld);
+    DeleteObject(hbm);
+    DeleteDC(hdcMem);
+
+    /* Finally draw the text over shadow */
+    crOldText = SetTextColor(hdc, crText);
+    SetBkMode(hdc, TRANSPARENT);
+    iRet = DrawTextW(hdc, pszText, cch, prc, dwFlags);
+    SetTextColor(hdc, crOldText);
 
-    FIXME("%p, %s, %d, %p, %#lx, %#lx, %#lx, %d, %d: semi-stub\n", hdc, debugstr_w(text),
-        length, rect, flags, crText, crShadow, offset_x, offset_y);
-
-    bkmode = SetBkMode(hdc, TRANSPARENT);
-    clr = SetTextColor(hdc, crShadow);
-
-    /* FIXME: for shadow we need to render normally, blur it, and blend with current background. */
-    r = *rect;
-    OffsetRect(&r, 1, 1);
-    DrawTextW(hdc, text, length, &r, flags);
-
-    SetTextColor(hdc, crText);
-
-    /* with text color on top of a shadow */
-    ret = DrawTextW(hdc, text, length, rect, flags);
-
-    SetTextColor(hdc, clr);
-    SetBkMode(hdc, bkmode);
-
-    return ret;
+    return iRet;
 }
 
 /***********************************************************************
--- a/src/dlls/comctl32/listview.c
+++ b/src/dlls/comctl32/listview.c
@@ -4752,7 +4752,16 @@
     if (infoPtr->uView == LV_VIEW_DETAILS && infoPtr->dwLvExStyle & LVS_EX_GRIDLINES)
         rcLabel.bottom--;
 
-    DrawTextW(nmlvcd->nmcd.hdc, item->pszText, -1, &rcLabel, format);
+    if ((!(item->state & LVIS_SELECTED) || !infoPtr->bFocus) &&
+        (infoPtr->dwLvExStyle & LVS_EX_TRANSPARENTSHADOWTEXT))
+    {
+        DrawShadowText(nmlvcd->nmcd.hdc, item->pszText, -1, &rcLabel, format,
+                       RGB(255, 255, 255), RGB(0, 0, 0), 2, 2);
+    }
+    else
+    {
+        DrawTextW(nmlvcd->nmcd.hdc, item->pszText, -1, &rcLabel, format);
+    }
 }
 
 /***
--- a/src/dlls/comctl32/toolbar.c
+++ b/src/dlls/comctl32/toolbar.c
@@ -612,6 +612,47 @@
               wine_dbgstr_rect(rcText));
 
 	hOldFont = SelectObject (hdc, infoPtr->hFont);
+
+    {
+        HTHEME theme = GetWindowTheme(infoPtr->hwndSelf);
+        if (theme && !(state & CDIS_DISABLED))
+        {
+            HWND hwndParent = GetParent(infoPtr->hwndSelf);
+            WCHAR className[32];
+            BOOL isTaskbar = FALSE;
+            if (hwndParent && GetClassNameW(hwndParent, className, ARRAY_SIZE(className)))
+            {
+                if (!lstrcmpiW(className, L"MSTaskSwWClass") ||
+                    !lstrcmpiW(className, L"Shell_TrayWnd"))
+                {
+                    isTaskbar = TRUE;
+                }
+            }
+
+            if (isTaskbar)
+            {
+                COLORREF textColor;
+                int stateId = TS_NORMAL;
+                if (state & CDIS_SELECTED)
+                    stateId = TS_PRESSED;
+                else if (state & CDIS_CHECKED)
+                    stateId = (state & CDIS_HOT) ? TS_HOTCHECKED : TS_HOT;
+                else if (state & CDIS_HOT)
+                    stateId = TS_HOT;
+
+                if (SUCCEEDED(GetThemeColor(theme, TP_BUTTON, stateId, TMT_TEXTCOLOR, &textColor)))
+                {
+                    int oldBk = SetBkMode(hdc, TRANSPARENT);
+                    DrawShadowText(hdc, lpText, -1, rcText, infoPtr->dwDTFlags,
+                                   textColor, RGB(0, 0, 0), 1, 1);
+                    SetBkMode(hdc, oldBk);
+                    SelectObject (hdc, hOldFont);
+                    return;
+                }
+            }
+        }
+    }
+
 	if ((state & CDIS_HOT) && (dwItemCDFlag & TBCDRF_HILITEHOTTRACK )) {
 	    clrOld = SetTextColor (hdc, tbcd->clrTextHighlight);
 	}
